### PR TITLE
matrices: remove the recursion on the qr decomposition

### PR DESCRIPTION
**Pull request by uLipe**
This is not the most elegant solution, but it still remove the recursion and makes the stack cost fixed.

### CODE REVIEW

**Comment by uLipe** [comment]
@microbuilder @gmarull  PTAL

**Comment by microbuilder** [comment]
@uLipe If you don't mind setting up a manifest PR in Zephyr against this, that will trigger a CI run for zscilib.

File: src/matrices.c [hunk 1]
```diff
@@ -1333,7 +1352,12 @@ zsl_mtx_qrd(struct zsl_mtx *m, struct zsl_mtx *q, struct zsl_mtx *r,
 		/* Reduce the matrix by 'g' rows and columns each time. */
 		ZSL_MATRIX_DEF(mred, (m->sz_rows - g), (m->sz_cols - g));
 		ZSL_MATRIX_DEF(hred, (m->sz_rows - g), (m->sz_rows - g));
-		zsl_mtx_reduce_iter(r, &mred);
+
+		/* allocatest the place holder matrices for the reduction loop */
```
Comment (microbuilder) [inline src/matrices.c:1356]: nit: `Allocate the place holder ...`

**Comment by uLipe** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/51041

cc: @microbuilder

**Review by microbuilder — APPROVED** [review]

**Comment by microbuilder** [comment]
> Tests fixed

Thanks, please update the Zephyr PR now that this is merged.

**Pull request merged**